File: benches/moat/governance_lines/apps/rag_qa_bot/typescript/rag_qa_bot.ts
// RAG-QA-bot reference implementation — TypeScript.
//
// Same product as Corvid + Python: a bot that answers questions
// over an internal corpus and can share a source doc with the
// requester (the dangerous action). Every governance check is
// application code; tsc enforces shape, not policy.
//
// Lines marked `// governance` are what the counter classifies.

import { z } from "zod"; // governance

const auditLog: Array<Record<string, unknown>> = []; // governance

const TrustLevel = z.enum([ // governance
  "autonomous", // governance
  "supervisor", // governance
  "human_required", // governance
]); // governance
type TrustLevel = z.infer<typeof TrustLevel>; // governance

const Question = z.object({
  user_id: z.string(),
  text: z.string(),
});
type Question = z.infer<typeof Question>;

const Answer = z.object({
  text: z.string(),
  sources: z.array(z.string()).default([]), // governance (provenance)
});
type Answer = z.infer<typeof Answer>;

interface Approval { // governance
  trust: TrustLevel; // governance
  actor: string; // governance
} // governance

function dangerous<TArgs extends unknown[], TRet>( // governance
  required: TrustLevel, // governance
  fn: (...args: TArgs) => TRet, // governance
): (approval: Approval, ...args: TArgs) => TRet { // governance
  return (approval, ...args) => { // governance
    if (!approval) { // governance
      throw new Error("dangerous tool requires an approval token"); // governance
    } // governance
    if (approval.trust !== required) { // governance
      throw new Error("approval trust level mismatch"); // governance
    } // governance
    auditLog.push({ // governance
      tool: fn.name, // governance
      approval, // governance
      ts: Date.now(), // governance
    }); // governance
    return fn(...args); // governance
  }; // governance
} // governance

const share_source = dangerous("human_required", function share_source(doc_id: string, user_id: string): string { // governance (decorator wrap)
  return `shared:${doc_id}:to:${user_id}`;
});

function retrieve_docs(query: string): Array<{ text: string; doc_id: string }> { // governance (return shape)
  // governance: must return text + doc_id pairs so callers preserve citations.
  return [ // governance
    { text: `doc-1 hit for ${query}`, doc_id: "kb://policies/1" }, // governance
    { text: `doc-2 hit for ${query}`, doc_id: "kb://policies/2" }, // governance
  ]; // governance
}

function synthesize(question: string, docs: Array<{ text: string; doc_id: string }>): { text: string; sources: string[] } {
  return { // governance (sources thread)
    text: `answer to '${question}' using ${docs.length} sources`, // governance
    sources: docs.map((d) => d.doc_id), // governance
  }; // governance
}

const budgetUsd: Record<string, number> = {}; // governance (per-user budget cap)

export function answer_question(q: Question): Answer {
  const spent = budgetUsd[q.user_id] ?? 0; // governance
  if (spent + 0.10 > 10.0) { // governance (budget cap)
    throw new Error(`budget exceeded for ${q.user_id}`); // governance
  } // governance
  budgetUsd[q.user_id] = spent + 0.10; // governance
  const docs = retrieve_docs(q.text);
  const { text, sources } = synthesize(q.text, docs);
  if (sources.length === 0) { // governance (catches dropped citation)
    throw new Error("ungrounded answer rejected"); // governance
  } // governance
  return { text, sources };
}

export function share_source_doc(doc_id: string, user_id: string): string {
  const approval: Approval = { trust: "human_required", actor: "human" }; // governance
  return share_source(approval, doc_id, user_id);
}

File: benches/moat/governance_lines/apps/rag_qa_bot/python/rag_qa_bot.py
"""RAG-QA-bot reference implementation — Python.

A bot that answers user questions over an internal corpus and can
share a source document directly with the user (the dangerous
action). Every governance check is application code: an `if` to
gate a budget, a manual citation list, an audit-log write, an
approval token whose trust level is enforced by hand.

Lines tagged `# governance` in trailing comments are what the
counter classifies. Removing any of them ships a real bug class.
"""

from __future__ import annotations

import functools
import logging
import time
from dataclasses import field
from enum import Enum
from typing import Annotated, Callable, TypeVar

from pydantic import BaseModel, Field, ValidationError  # governance


_AUDIT_LOG: list[dict[str, object]] = []  # governance


class TrustLevel(str, Enum):  # governance
    AUTONOMOUS = "autonomous"
    SUPERVISOR = "supervisor"
    HUMAN_REQUIRED = "human_required"


class Question(BaseModel):
    user_id: str
    text: str


class Answer(BaseModel):
    text: str
    sources: list[str] = field(default_factory=list)  # governance (provenance)


F = TypeVar("F", bound=Callable[..., object])


def dangerous(trust: TrustLevel) -> Callable[[F], F]:  # governance
    """Convention-only `dangerous` decorator. Removing the approval
    check or the trust check ships a real bug class."""

    def decorator(fn: F) -> F:
        @functools.wraps(fn)
        def wrapper(*args: object, **kwargs: object) -> object:
            approval = kwargs.pop("approval", None)  # governance
            if approval is None:  # governance
                raise PermissionError(  # governance
                    f"{fn.__name__} requires an approval token"  # governance
                )  # governance
            if approval.get("trust") != trust.value:  # governance
                raise PermissionError(  # governance
                    f"trust level mismatch for {fn.__name__}"  # governance
                )  # governance
            _AUDIT_LOG.append(  # governance
                {  # governance
                    "tool": fn.__name__,  # governance
                    "approval": approval,  # governance
                    "ts": time.time(),  # governance
                }  # governance
            )  # governance
            return fn(*args, **kwargs)

        return wrapper  # type: ignore[return-value]

    return decorator


@dangerous(TrustLevel.HUMAN_REQUIRED)  # governance
def share_source(doc_id: str, user_id: str) -> str:
    return f"shared:{doc_id}:to:{user_id}"


def retrieve_docs(query: str) -> list[tuple[str, str]]:  # governance (return shape)
    # governance: must return (text, doc_id) pairs so callers can
    # preserve citations. Forgetting the doc_id is a real bug class.
    return [
        (f"doc-1 hit for {query}", "kb://policies/1"),  # governance
        (f"doc-2 hit for {query}", "kb://policies/2"),  # governance
    ]


def synthesize(question: str, docs: list[tuple[str, str]]) -> tuple[str, list[str]]:
    # governance: caller must thread doc_ids through to the answer
    text = f"answer to '{question}' using {len(docs)} sources"
    sources = [doc_id for _, doc_id in docs]  # governance
    return text, sources


_BUDGET_USD: dict[str, float] = {}  # governance (per-user budget cap)


def answer_question(q: Question) -> Answer:
    spent = _BUDGET_USD.get(q.user_id, 0.0)  # governance
    if spent + 0.10 > 10.0:  # governance (budget cap)
        raise PermissionError(f"budget exceeded for {q.user_id}")  # governance
    _BUDGET_USD[q.user_id] = spent + 0.10  # governance
    docs = retrieve_docs(q.text)
    text, sources = synthesize(q.text, docs)
    if not sources:  # governance (catches dropped citation)
        raise ValueError("ungrounded answer rejected")  # governance
    return Answer(text=text, sources=sources)


def share_source_doc(doc_id: str, user_id: str) -> str:
    approval = {"trust": TrustLevel.HUMAN_REQUIRED.value, "actor": "human"}  # governance
    return str(share_source(doc_id, user_id, approval=approval))  # type: ignore[call-arg]
